### PR TITLE
[WIP] V2 low-level operator implementation

### LINKED ISSUE #619: Develop the workspace API and controller
Implementation task for the `Workspace` API, per [design doc](https://docs.google.com/document/d/1hHOgs9uDsvAbj89D-b4WjVG9GnoJ7_cy5FKGOeSAfRg/edit#heading=h.o1cg2qaija75).

### PR DESCRIPTION
**Pull request by EronWright**
### Proposed changes

This PR implements the `auto.pulumi.com` API Group, including the `Workspace` and `Update` types.

Integration tests for the workspace controller are included.  

### Example
```yaml
apiVersion: auto.pulumi.com/v1alpha1
kind: Workspace
metadata:
  name: random-yaml-1e2fc47
spec:
  image: pulumi/pulumi:3.128.0-nonroot
  securityProfile: restricted  
  serviceAccountName: default
  git:
    url: https://github.com/pulumi/examples.git
    revision: 1e2fc471709448f3c9f7a250f28f1eafcde7017b
    dir: random-yaml
  env:
  - name: PULUMI_ACCESS_TOKEN
    valueFrom:
      secretKeyRef:
        name: pulumi-api-secret
        key: accessToken
  resources:
    requests:
      cpu: 1
      memory: 512Mi
    limits:
      cpu: 1
      memory: 512Mi

  # various extension points shown here.
  # - custom pod labels
  # - pod tolerations
  # - extra init container(s)
  # - extra volume(s) and volume mounts onto the 'pulumi' container
  podTemplate:
    metadata:
      labels:
        example.com/mylabel: bar
    spec:
      terminationGracePeriodSeconds: 3600
      tolerations:
        - key: "example.com/foo"
          operator: "Exists"
          effect: "NoSchedule"
      initContainers:
      - name: extra
        image: busybox
        command: ["sh", "-c", "echo 'Hello, extra init container!'"]
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            add:
            - NET_BIND_SERVICE
            drop:
            - ALL
        volumeMounts:
          - name: share
            mountPath: /share
      containers:
      - name: pulumi
        volumeMounts:
          - name: secret-volume
            mountPath: /etc/secret-volume
            readOnly: true
      volumes:
        - name: secret-volume
          secret:
            secretName: test-secret
```

### Specific Changes

- (agent) git source support
- (operator) scaffolding
- (operator) workspace api and controller and tests
- (operator) update api and controller
- (experimental) configuration block in `Environment` spec 
- (experimental) dockerfile for non-root pu/pu

### Related issues (optional)

Closes #619

### CODE REVIEW

File: operator/test/utils/utils.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+)
+
+func warnError(err error) {
+	fmt.Fprintf(GinkgoWriter, "warning: %v\n", err)
+}
+
+// InstallPrometheusOperator installs the prometheus Operator to be used to export the enabled metrics.
+func InstallPrometheusOperator() error {
+	url := fmt.Sprintf(prometheusOperatorURL, prometheusOperatorVersion)
+	cmd := exec.Command("kubectl", "create", "-f", url)
```
[the remark this answers is not in the document]
Comment (EronWright) [reply operator/test/utils/utils.go:44]: This code is assumedly used by the e2e test that was scaffolded but I haven't really looked at. Can we ignore it for now?

File: operator/.golangci.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  # (fill in the rest as needed)
+  exclude-rules:
+    - path: "api/*"
+      linters:
+        - lll
+    - path: "internal/*"
+      linters:
+        - dupl
+        - lll
+linters:
```
Comment (blampe) [inline operator/.golangci.yml:19]: Can we keep this consistent with the linter's we're using in p-k?

**Review by blampe — APPROVED** [review]
Nothing blocking!

File: agent/cmd/init.go [hunk 3]
```diff
@@ -53,25 +61,78 @@ For Flux sources:
...
+		// fetch the configured git artifact
+		if GitUrl != "" {
+			u, err := url.Parse(GitUrl)
+			if err != nil {
+				log.Errorw("fatal: unable to parse git url", zap.Error(err))
+				os.Exit(2)
+			}
+			// Configure authentication strategy to access the source
+			authData := map[string][]byte{}
+			authOpts, err := git.NewAuthOptions(*u, authData)
```
[the remark this answers is not in the document]
Comment (EronWright) [reply agent/cmd/init.go:90]: I intend to support the same options, but I was waiting to see how containerization would impact the options.  e.g., the ability to mount the git credentials.

File: operator/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+COPY api/ api/
+COPY agent/ agent/
+COPY internal/controller/ internal/controller/
+
+# Build
+# the GOARCH has not a default value to allow the binary be built according to the host where the command
+# was called. For example, if we call make docker-build in a local env which has the Apple Silicon M1 SO
+# the docker BUILDPLATFORM arg will be linux/arm64 when for Apple x86 it will be linux/amd64. Therefore,
+# by leaving it empty we can ensure that the container and binary shipped on it will have the same platform.
+RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o manager cmd/main.go
```
[the remark this answers is not in the document]
Comment (EronWright) [reply operator/Dockerfile:28]: Yes it is entirely boilerplate from kubebuilder v.latest.

**Pull request merged**